### PR TITLE
FCD from fcd_torch

### PR DESCRIPTION
**Pull request by danpol**
1. Uses faster FCD implementation from fcd_torch
2. Removed all dependencies from Keras/Tensorflow in package
3. flake8 for metrics for further CI code style tests
4. Removed all dependencies from Keras/Tensorflow from Dockerfile
5. Save config and vocab before training to have functional checkpoints if training crashes

### CODE REVIEW

**Comment by danpol** [comment]
Resolves #28

File: moses/metrics/metrics.py [hunk 1]
```diff
@@ -1,37 +1,39 @@
...
         batch_size: batch size for FCD metric
         test_scaffolds: list of scaffold test SMILES
             Will compute only on the general test set if not specified
         ptest: dict with precalculated statistics of the test set
-        ptest_scaffolds: dict with precalculated statistics of the scaffold test set
-        
-    
+        ptest_scaffolds: dict with precalculated statistics
+            of the scaffold test set
+        gpu: deprecated, use `device`
```
Comment (shayakhmetov) [inline moses/metrics/metrics.py]: Если gpu уже удалили, то и в описании наверное надо удалить

File: moses/metrics/metrics.py [hunk 2]
```diff
@@ -40,73 +42,99 @@ def get_all_metrics(test, gen, k=[1000, 10000], n_jobs=1, gpu=-1,
...
+    """
+    if k is None:
+        k = [1000, 10000]
     disable_rdkit_log()
     metrics = {}
-    if n_jobs != 1:
-        pool = Pool(n_jobs)
-    else:
-        pool = 1
+    if gpu is not None:
```
Comment (shayakhmetov) [inline moses/metrics/metrics.py]: gpu убрали ведь?

File: moses/metrics/metrics.py [hunk 3]
```diff
@@ -40,73 +42,99 @@ def get_all_metrics(test, gen, k=[1000, 10000], n_jobs=1, gpu=-1,
...
     statistics = {}
     mols = mapper(pool)(get_mol, smiles)
-    kwargs = {'n_jobs': n_jobs, 'gpu': gpu, 'batch_size': batch_size}
-    statistics['FCD'] = FCDMetric(**kwargs).precalc(smiles)
+    kwargs = {'n_jobs': pool, 'device': device, 'batch_size': batch_size}
+    kwargs_fcd = {'n_jobs': n_jobs, 'device': device, 'batch_size': batch_size}
+    statistics['FCD'] = FCDMetric(**kwargs_fcd).precalc(smiles)
     statistics['SNN'] = SNNMetric(**kwargs).precalc(mols)
-    statistics['Frag'] = FragMetric(**kwargs).precalc(mols)
+    statistics['Frag'] = FragMetric(**kwargs_fcd).precalc(mols)
```
Comment (shayakhmetov) [inline moses/metrics/metrics.py]: здесь точно kwargs_fcd?

File: moses/metrics/metrics.py [hunk 4]
```diff
@@ -181,107 +215,90 @@ def remove_invalid(gen, canonize=True, n_jobs=1):
 
 
 class Metric:
-    def __init__(self, n_jobs=1, gpu=-1, batch_size=512, **kwargs):
+    def __init__(self, n_jobs=1, device=-1, batch_size=512, **kwargs):
```
Comment (shayakhmetov) [inline moses/metrics/metrics.py]: device=-1 норм? Может device='cpu'?

**Pull request merged**